README
Timing tests with respect to PNET module.

I don't have a spec that tells me anything
about the phase relationship of the 360Hz
interrupt trigger (VMTG) with respect to
the fiducial. So the best we can do ATM is
measuring the VMTG IRQ phase with respect
to the PNET interrupt and then adjust the
VMTG delay so that we get the usual
150us post PNET arrival.

a)  CSR0_IRG = 0 (falling edge of 360Hz signal)
    INTIPRSC = 0
    INTIDLY  = 0
   vmtgIRQTime - pnetIRQTime = 2.36ms

b) set prescaler to 80 so the delay is conveniently
   in microseconds.
    INTIPRSC = 80
    INTIDLY  = 80
   -> still 2.36ms as expected

c) set delay to 1/360Hz - 2360us + 150us ~= 570us
    INTIPRSC = 80
    INTIDLY  = 570
   vmtgIRQTime - pnetIRQTime = 158us

Then also adjust TSZDDLY so that PNET and VMTG
timeslot numbers match. The VMTG's TS counter
increments synchronously with the VMTG interrupt.
Use VMTG_TSDIFF PV to read offset from PNET to
VMTG interrupt.


RELEASE_NOTES
R2.0.1:      (Kukhee Kim, 2021/01/11)
  - add vmtgConfigure() command to allow to initialize driver explicitly with manual order 

R2.0.0:      (E. Williams, 2020-Oct-01)
             This module was converted to GIT and migrated to EPICS 7.
             EPICS 7 is a major release of EPICS.  However, there is no functional
             change in this module as a result of the upgrade to EPICS 7.

VMTG-R1-4-0  (C. Bianchini,  2019/11/04) 
 - Compiled with Autosave
 - Added save macro field to VMTGlirw.template, VMTGPerFrq.template, VMTGbo.template, VMTGAdcAi.template, VMTGAdcAo.template
   and resolved in VMTG.substitutions

VMTG-R1-3-1  (C. Bianchini,  2018/12/12)
 - Added VMTGReSyncFrqAvg.db to calculate average sync frequency every 10 counts.
 - Added CALC record to move Hz to MHz in VMTGApp/Db/VMTGReSyncFrqAvg.db .

VMTG-R1-3-0  (C. Bianchini, 2018/11/29)
 - Added new fields to VMTG.substitutions
  Changes for 0x00A2- 0x00BA:
  file VMTGPerFrq.template {
    pattern { field        , off    , lopr    , hopr    , lolo    , low     , high    , hihi    , llsv       , lsv        , hsv        , hhsv       , desc            }
          { "SyncPER"    , "0xBA" , "0"     , "1000"  , "0"     , "0"     , "0"     , "0"     , "NO_ALARM" , "NO_ALARM" , "NO_ALARM" , "NO_ALARM" , "Internal 8.5MHz Resync Clk"}
  }

  Changes for 0x0096:
  file VMTGStatusBit.template {
    pattern { field        , off    , boio         , mask    , znam          , onam              , high , desc                }
              { "RCLKSEL_GCSL","0x96" , "vmtgCSR2Lo" , "0x0001", "Int 8.5MHz Clk Sel", "Out ECSL mux Sel", "0" , "Global Resync Clk Sel"}
              { "RCLKSEL_CDCS","0x96" , "vmtgCSR2Lo" , "0x0100", "Int default param", "custom param", "0" , "VMTG Trig Cnt Logic Decode Param Sel"}
  And: (commented out and used the following template)
  file VMTGmbbo.template {
    pattern { field        , off    , shft    , nobt , zrst            , onst            , twst            , thst            , frst            , fvst            , sxst            , svst            , desc    
           { "RCLKSEL_ECSL"  , "0x96" , "12"    , "2"  , "Level"         , "Internal div Resync Clk" , "Ext Resync Clk src" , "Ext Resync Clk w 119MHz Clk"    , ""     , ""              , ""              , ""              , "Ext Resync Clk Sel"  }

VMTG-R1-2-2  (S. Allison, 2016/06/27)
 - Upgrade to devBusMapped-R3-1-1 and miscUtils-R2-2-3.

VMTG-R1-2-1  (C. Bianchini, 2016/01/13)
 - Add VMTG flag to identify module's presents, it is used by Mpg app.

VMTG-R1-2-0 (K. Kim, 2014/03/31):
 - Add clock re-synchronize register in the VMTG register

VMTG-R1-1-1 (S. Allison, 2013/05/10): Use R3-14-12. 

VMTG-R1-1-0 (T. Straumann, 2013/04/26): Bugfix release 

 - fixed typo; the outlinks in the F02 and F03 fanouts should be
   to ADC_Cx_LO, ADC_Cy_HI (not ADC_CHx_LO/HI). Thanks to Stephanie
   Allison for catching this.

 - devBusMapped-R3-1-0 has changed API for I/O methods. Need
   to adapt!

VMTG-R1-0-1: Use a newer devBusMapped: devBusMapped-R3-1-0


2011/10/28: First 'stable' release VMTG-R1-0-0 released.

*** NOTE **********************************************************************
*
*  We need a reasonably recent epics base version (devAllocAddr doesn't deliver
*  requested alignment under 3.14.8.2; 3.14.12 works but it is possible that
*  an intermediate version is good, too).
*
*******************************************************************************


VMTGApp/src/drvVMTG.h
#ifndef DRV_VMTG_H
#define DRV_VMTG_H
/* $Id: drvVMTG.h,v 1.1 2011/06/01 17:10:06 strauman Exp $ */

/* VMTG Driver public interface */

#ifdef __cplusplus
extern "C" {
#endif

/* Interrupt bit definitions; note that the VME IRQ level is
 * encoded in the upper bits.
 */

#define REG_IRQ_CTRL_LVL_MASK    0xfe00
#define REG_IRQ_CTRL_LVL(l)      (1<<((l)+8))

#define REG_IRQ_CTRL_EI          (1<<0)
#define REG_IRQ_CTRL_TS0I        (1<<1)
#define REG_IRQ_CTRL_DRI         (1<<2)
#define REG_IRQ_CTRL_TMI         (1<<3)
#define REG_IRQ_CTRL_MS0I        (1<<4)
#define REG_IRQ_CTRL_GTI         (1<<5)
#define REG_IRQ_CTRL_C8MI        (1<<6)
#define REG_IRQ_CTRL_RFMI        (1<<7)
#define REG_IRQ_CTRL_GPI         (1<<8)


/* ISR. The routine is passed a bit-set of 'pending' interrupts
 * and it must return a bitmask of the 'handled' interrupt conditions.
 * 'parm' is the parameter passed to the installation routine.
 */
typedef uint16_t (*VMTGIsr)(uint16_t irqs_pending, void *parm);

/* Install an ISR; RETURNS zero on success, nonzero on error */

long
vmtgInstallISR(VMTGIsr isr, void *parm);

/* Disable interrupts at the VMTG board.
 * RETURNS: bitmask to pass to vmtgIrqsEnable().
 * NOTE:    disable/enable pairs can be nested
 *          and only the outermost 'enable' will
 *          effectively re-enable the interrupts.
 */
uint16_t
vmtgIrqsDisable(void);

/*
 * Enable interrupts. 'key' is a bitmask returned
 * previously by vmtgIrqsDisable().
 *
 * The first time 'vmtgIrqsEnable()' is executed
 * the key must be computed by the user. It is
 * composed of a bit-mask of interrupt conditions
 * AND an (encoded) interrupt level.
 *
 * NOTE: This routine enables interrupts only
 *       at the board. The user must ensure
 *       that the corresponding level is enabled
 *       at the VME controller.
 */
void
vmtgIrqsEnable(uint16_t key);

/* Obtain current time-slot
 * RETURNS:  one-based time-slot, i.e., a number [1..6]
 */
int
vmtgGetTS(void);

/* Obtain current time-slot MODULO 3
 * RETURNS:  one-based time-slot MOD 3, i.e., a number [1..3]
 */
int
vmtgGetTSMod3(void);

#ifdef __cplusplus
}
#endif

#endif

#define HAVE_VMTG 1

#define NO_VMTG 0


int

getVmtg_flag(void);



VMTGApp/src/vme64xSup/vme64x.c
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

#include <vme64x.h>

#define DEBUG

/* read a 24-bit register composed of 3 bytes spaced by 4 */
uint32_t vme64CSRegRead24(VME64_Addr base, uint32_t off)
{
	return __vme64CSRegRead24(base,off);
}


/* read a 32-bit register composed of 4 bytes spaced by 4 */
uint32_t vme64CSRegRead32(VME64_Addr base, uint32_t off)
{
	return __vme64CSRegRead32(base,off);
}

/* read a 64-bit register composed of 8 bytes spaced by 4 */
uint64_t vme64CSRegRead64(VME64_Addr base, uint32_t off)
{
	return __vme64CSRegRead64(base,off);
}

/* write a 32-bit register composed of 4 bytes spaced by 4 */
void vme64CSRegWrite32(VME64_Addr base, uint32_t off, uint32_t val)
{
	__vme64CSRegWrite32(base, off, val);
}

uint32_t
vme64ReadADER(VME64_Addr base, uint32_t idx)
{
	return vme64CSRegRead32(base, VME64_CSR_OFF_ADER(idx));
}

void
vme64WriteADER(VME64_Addr base, uint32_t idx, uint32_t val)
{
	return vme64CSRegWrite32(base, VME64_CSR_OFF_ADER(idx), val);
}

uint32_t
vme64ReadADEM(VME64_Addr base, uint32_t idx)
{
	return vme64CSRegRead32(base, VME64_CR_OFF_ADEM(idx));
}

uint64_t
vme64ReadAMCAP(VME64_Addr base, uint32_t idx)
{
	return vme64CSRegRead64(base, VME64_CR_OFF_AMCAP(idx));
}

#ifdef DEBUG
int
vme64PrintAMCAP(VME64_Addr base, uint32_t idx)
{
	printf("AMCAP(%"PRIu32"): 0x%016"PRIx64"\n", idx, vme64ReadAMCAP(base,idx));
	return 0;
}
#endif

int
vme64SetupADER(VME64_Addr base, uint32_t idx, uint64_t addr, uint32_t am)
{
uint64_t cap;
uint64_t msk;
uint32_t adem;

	if (idx < 0 || idx > 7 ) {
		fprintf(stderr,"vme64SetupADER(): index %"PRIu32" out of range\n", idx);
		return -1;
	}

	if ( am > 63 ) {
		fprintf(stderr,"vme64SetupADER(): invalid AM requested\n");
		return -1;
	}

	cap = vme64ReadAMCAP(base,idx);
	if ( ! (cap & (1ULL<<am)) ) {
		fprintf(stderr,"vme64SetupADER(): requested AM 0x%02"PRIx32" not supported by function %"PRIu32"\n", am, idx);
		return -1;
	}

	adem = vme64ReadADEM(base,idx);
	msk  = adem & ~0xff;

	if ( adem & VME64_ADEM_EFM ) {
		fprintf(stderr,"vme64SetupADER(): Extra function mask not supported\n");
		return -1;
	}
	if ( adem & VME64_ADEM_EFD ) {
		fprintf(stderr,"vme64SetupADER(): Extra function decoder not supported\n");
		return -1;
	}
	if ( adem & VME64_ADEM_DFS ) {
		fprintf(stderr,"vme64SetupADER(): Dynamic function sizing not supported\n");
		return -1;
	}

	if ( ~msk & addr ) {
		fprintf(stderr,"vme64SetupADER(): requested address (0x%016"PRIx64")\n", addr);
		fprintf(stderr,"                  not on MASK       (0x%016"PRIx64") boundary\n", msk);
		return -1;
	}
	if ( adem & VME64_ADEM_FAF ) {
		if ( (vme64ReadADER(base, idx) & msk) != addr ) {
			fprintf(stderr,"vme64SetupADER(): Dynamic function sizing not supported\n");
			return -1;
		} else {
			return 0;
		}
	}

	vme64WriteADER(base, idx, addr | (am<<2));

	return 0;
}

uint32_t
vme64CRChecksum(VME64_Addr base, int quiet)
{
uint32_t off;
uint8_t  sum;
uint32_t len = 0;
uint8_t  w,sp;

	if (   'C' != vme64_in08(base, VME64_CR_OFF_C_ID)
		|| 'R' != vme64_in08(base, VME64_CR_OFF_R_ID) ) {
		if ( !quiet )
			fprintf(stderr,"'CR' ID not found -- is there a VME64 module at 0x%08"myPRIxPTR" ?\n", base);
		return -1;
	}

	len = vme64_in08(base, VME64_CR_OFF_LEN2);
	if ( len > 0x7 ) {
		if ( !quiet )
			fprintf(stderr,"CR Length MSB 0x%02"PRIx32" > 0x07", len);
		return -1;
	}
	len = (len<<8) | vme64_in08(base, VME64_CR_OFF_LEN1);
	len = (len<<8) | vme64_in08(base, VME64_CR_OFF_LEN0);

	for ( sum=0, off = VME64_CR_OFF_LEN2; len && off <= 0x7f ; off+=4, len--)
		sum += vme64_in08(base, off);

	/* Hmm - the standard says that e.g., for D32 access width every
	 * byte of the ROM is used. I would interpret that literally and
	 * count the (zero byte-0..byte2) locations into the ROM length
	 * but at least the VME digitizer does not.
	 */
	switch ( ( w = vme64_in08(base, VME64_CR_OFF_CRWD) ) ) {
		case VME64_CR_CRWD_D8_O:
			sp  = 4;
		break;

		case VME64_CR_CRWD_D8_EO:
			/* see above sp  = 2; */
			sp = 4;
		break;

		case VME64_CR_CRWD_D16:
		case VME64_CR_CRWD_D32:
			/* see above sp  = 1; */
			sp  = 4;
		break;

		default:
			fprintf(stderr,"Unsupported CR data access width 0x%02x"PRIx8"\n",w);
			return -1;
	}
	off = 0x7f + sp;
	while ( len ) {
		sum += vme64_in08(base, off);
		off += sp;
		len--;
	}
	return sum - vme64_in08(base,3);
}

/* Read serial-number info */
uint32_t
vme64ReadSN(VME64_Addr base)
{
VME64_Addr b,e;
int        sp   = -1;
uint32_t   rval = 0;
	b = base + __vme64CSRegRead24(base, VME64_CR_OFF_BEG_SN);
	e = base + __vme64CSRegRead24(base, VME64_CR_OFF_END_SN);
	/* Standard says that S/N should be on D08(O) - every fourth
	 * byte (address ends on 3,7,B or F).
	 * Use some heuristics if this rule is violated.
	 */
	if ( 3 == ( b & 3 ) && 3 == ( e & 3 ) && (int)(e-b) <= 3*4 ) {
		/* Ordinary case with S/N <= 4 bytes */
		sp = 4;
	} else if ( e-b <= 3 ) {
		/* They violate the rule but S/N is <= 4 bytes
		 * sitting there...
		 */
		sp = 1;
	}

	if ( -1 == sp ) {
		return (uint32_t)-1;
	}

	while ( b <= e ) {
		rval = (rval << 8) + vme64_in08(b, 0);
		b+=sp;
	}

	return rval;
}


int
vme64FindBoard(VME64_Addr base, int start_slot, uint32_t man_id, uint32_t brd_id)
{

	if ( start_slot < 1 )
		start_slot = 1;

	base += start_slot * VME64_CR_SPACING;

	while ( start_slot <= 21 ) {
		if (   'C' == vme64_in08(base, VME64_CR_OFF_C_ID)
			&& 'R' == vme64_in08(base, VME64_CR_OFF_R_ID) ) {
			if (    man_id == __vme64CSRegRead24(base, VME64_CR_OFF_MANID)
				 && brd_id == __vme64CSRegRead32(base, VME64_CR_OFF_BRDID) )
				return start_slot;
		}
		base += VME64_CR_SPACING;
		start_slot++;
	}

	return -1;
}


VMTGApp/src/drvVMTG.c
/* $Id: drvVMTG.c,v 1.10 2013/04/26 18:34:54 strauman Exp $ */

/* VMTG driver */

#include <epicsExport.h>
#include <epicsInterrupt.h>
#include <devLib.h>
#include <vme64x.h>
#include <drvSup.h>
#include <devBusMapped.h>
#include <errlog.h>
#include <cantProceed.h>
#include <inttypes.h>
#include <time.h>

#include <basicIoOps.h>

#include <drvVMTG.h>

#define  VMTG_MAN_ID	(('S'<<16) | ('L' << 8) | 'A')
#define  VMTG_BRD_ID    0x14407802

#define  VMTG_FUN       0 /* VMTG is function 0 */

#define  R32_HI_OFF     0
#define  R32_LO_OFF     2

#define REG_INTVEC      0x06
#define REG_INT_STAT    0x08
#define REG_IRQ_CTRL    0x0a
#define REG_TSCNT       0x36

#define TSCNT_M3(cnt)   ( ((cnt) >> 0 ) & 0x03 )
#define TSCNT_M6(cnt)   ( ((cnt) >> 4 ) & 0x07 )
#define TSCNT_M360(cnt) ( ((cnt) >> 7 ) & 0x1f )

#define MAX_VEC_RETRIES 10

static VMTGIsr  vmtgIsr = 0;

uint16_t       vmtgLfsr = 0xace1;

unsigned       vmtgVect = 0;
int            vmtgSlot = 0;
VME64_Addr     vmtgCSRB = (VME64_Addr)0;
volatile void *vmtgBase = (volatile void*)0;

static volatile int theVmtgFlag =HAVE_VMTG;

static int
vmtg32Rd(DevBusMappedPvt pvt, epicsUInt32 *pvalue, int idx, dbCommon *prec)
{
	idx    <<= 1;
	*pvalue  = 0;
 	*pvalue |= (in_be16( pvt->addr + idx + R32_HI_OFF ) << 16);
	*pvalue |= (in_be16( pvt->addr + idx + R32_LO_OFF ) & 0xffff);
	return 0;
}

static DevBusMappedAccessRec vmtg32IO = {
	rd: vmtg32Rd,
	wr: 0
};

static int
vmtgCSR2Rd(DevBusMappedPvt pvt, epicsUInt32 *pvalue, int idx, dbCommon *prec)
{
	if ( idx != 0 )
		cantProceed("Programming error - vmtgCSR2Rd can not be used with an idx != 0\n");
	*pvalue = in_be16( pvt->addr );
	return 0;
}

static int
vmtgCSR2WrLo(DevBusMappedPvt pvt, epicsUInt32 value, int idx, dbCommon *prec)
{
epicsUInt16 oval;
	/* When writing low bits (write-1-to-clear) then we don't want to
	 * touch the high bits.
	 */
	if ( idx != 0 )
		cantProceed("Programming error - vmtgCSR2WrLo can not be used with an idx != 0\n");
	oval  = in_be16( pvt->addr );
	oval  = (oval & 0xff00) | ( value & 0xff);
	out_be16( pvt->addr, oval );
	return 0;
}

static int
vmtgCSR2WrHi(DevBusMappedPvt pvt, epicsUInt32 value, int idx, dbCommon *prec)
{
	if ( idx != 0 )
		cantProceed("Programming error - vmtgCSR2WrHi can not be used with an idx != 0\n");
	/* When writing hi bits ('normal' bits) then we don't want to
	 * set any of the low bits.
	 */
	out_be16( pvt->addr, (value & 0xff00) );
	return 0;
}

static DevBusMappedAccessRec vmtgCSR2Lo = {
	rd: vmtgCSR2Rd,
	wr: vmtgCSR2WrLo
};

static DevBusMappedAccessRec vmtgCSR2Hi = {
	rd: vmtgCSR2Rd,
	wr: vmtgCSR2WrHi
};

static uint16_t
vmtgRnd()
{
/* Primitive Galois LFSR (Wikipedia, 2011) */
uint16_t rval;
int      key;
	key = epicsInterruptLock();
		rval = vmtgLfsr = (vmtgLfsr>>1) ^ ( (- (vmtgLfsr&1)) & 0xb400);
	epicsInterruptUnlock( key );
	return rval;
}

static void
vme_isr(void *parm)
{
uint16_t handled;

	handled = vmtgIsr( in_be16( vmtgBase + REG_INT_STAT ), parm );

	out_be16( vmtgBase + REG_INT_STAT, handled );
}

uint16_t
vmtgIrqsDisable(void)
{
int      k;
uint16_t rval;
	k = epicsInterruptLock();
		rval = in_be16( vmtgBase + REG_IRQ_CTRL );
		out_be16( vmtgBase + REG_IRQ_CTRL, rval & ~REG_IRQ_CTRL_LVL_MASK );
	epicsInterruptUnlock( k );

	return rval;
}

void
vmtgIrqsEnable(uint16_t key)
{
	out_be16( vmtgBase + REG_IRQ_CTRL, key );
}

int
vmtgGetTS(void)
{
	return TSCNT_M6( in_be16( vmtgBase + REG_TSCNT ) ) + 1;
}

int
vmtgGetTSMod3(void)
{
	return TSCNT_M3( in_be16( vmtgBase + REG_TSCNT ) ) + 1;
}

long
vmtgInstallISR(VMTGIsr isr, void *parm)
{
int      i;
unsigned v;
long     rval;

	if ( ! isr || vmtgIsr )
		return -1;

	i = 0;

	do {
		v    = vmtgRnd() & 0xff;
		rval = devConnectInterruptVME( v, vme_isr, parm );
	} while ( S_dev_vectorInUse == rval && i++ < MAX_VEC_RETRIES );

	if ( S_dev_success == rval ) {
		vmtgVect = v;
		vmtgIsr  = isr;
		out_be16( vmtgBase + REG_INTVEC, v );
	}
	
	return rval; 

}

static uint32_t nsdiff(uint32_t b, uint32_t e)
{
	if ( e < b )
		e += 1000000000L;
	return e-b;
}

volatile uint32_t vmtgTestIRQT0PR =  0;
volatile uint32_t vmtgTestIRQTSPR =  0;
volatile uint32_t vmtgTestIRQDiff = -1;
volatile uint32_t vmtgTestIRQCntr =  0;

/* This ISR measures the time elapsed between a TS0 interrupt
 * and an 'external' interrupt occurring on TS0. This delay should
 * be the difference between the (programmable) TS0 and EI delays.
 */
uint16_t
vmtgTestISR(uint16_t irqs_pending, void *uarg)
{
static uint32_t start_t0;
static uint32_t start_ts;

struct timespec now;

	vmtgTestIRQCntr++;

	/* filter-out interrupts we're interested in */
	irqs_pending &= (REG_IRQ_CTRL_EI | REG_IRQ_CTRL_TS0I);

	clock_gettime( CLOCK_REALTIME, &now );

	if ( irqs_pending == (REG_IRQ_CTRL_EI | REG_IRQ_CTRL_TS0I) ) {
		/* Not enough delay between the two interrupts */
		vmtgTestIRQDiff = 0;
		start_t0 = start_ts = now.tv_nsec;
	} else if ( (REG_IRQ_CTRL_EI & irqs_pending) ) {
		if ( 1 == vmtgGetTS() ) {
			vmtgTestIRQDiff = nsdiff( start_t0, now.tv_nsec );
			vmtgTestIRQTSPR = nsdiff( start_ts, now.tv_nsec );
			start_ts        = now.tv_nsec;
		}
	} else if ( (REG_IRQ_CTRL_TS0I & irqs_pending) ) {
		vmtgTestIRQT0PR = nsdiff( start_t0, now.tv_nsec );
		start_t0        = now.tv_nsec;
	}

	return irqs_pending;
}


static long
vmtgReport(int interest_level)
{
	epicsPrintf("VMTG Driver, Release $Name:  $\n");
	if ( vmtgSlot < 1 ) {
		epicsPrintf("No VMTG card detected\n");
	} else {
		epicsPrintf("VMTG in slot #%i, @0x%"PRIxVME64A" (CSR @0x%"PRIxPTR"), IRQ Vector: %u\n",
		            vmtgSlot, (uintptr_t)vmtgBase, vmtgCSRB, vmtgVect);
	}
	return 0;
}

static long
vmtgInit(void)
{
static   uint32_t already = 0;
volatile void *vme_csr_base, *vme_24_base, *tmp;
uint32_t  vme_24_bus;
long      rval = -1;
uint32_t  adem;
int       lsb;

if(already) return 0;          /*  it has been intialized */
else        already = 1;       /*  it needs to be processed */

if (HAVE_VMTG==theVmtgFlag){ 
	/* Look-up CPU address of VME CSR space */
	if ( devBusToLocalAddr( atVMECSR, 0, &vme_csr_base) ) {
		epicsPrintf("vmtgInit FAILED: Unable to find VME CSR space of this board; cannot initialize VMTG driver\n");
		goto bail;
	}

	/* Scan VME64 for a board */
	vmtgSlot = vme64FindBoard( (VME64_Addr)vme_csr_base, 0, VMTG_MAN_ID, VMTG_BRD_ID );
	if ( vmtgSlot < 0 ) {
		epicsPrintf("vmtgInit FAILED: No VMTG Card found!\n");
		goto bail;
	}

	vmtgCSRB = (VME64_Addr)vme_csr_base + vmtgSlot * VME64_CR_SPACING;

	/* How much space do we need ? */
	adem = vme64ReadADEM( vmtgCSRB, VMTG_FUN );

	for ( lsb=8, adem>>=8; 0 == (adem & 1); lsb++ )
		adem >>= 1;

	/* Allocate a chunk of A24 addresses for this card */
	if ( devAllocAddress( "VMTG", atVMEA24, (1<<lsb), lsb, &vme_24_base ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to map %u bytes of VME A24 space\n", (1<<lsb));
		goto bail;
	}

	vmtgBase = vme_24_base;

	/* Hmm - there is no devLocalToBusAddr() but we need the *bus* address
	 * for programming the ADER.
	 */

	if ( devBusToLocalAddr( atVMEA24, 0, &tmp ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to map VME A24(0)\n");
		goto bail;
	}

	vme_24_bus = vme_24_base - tmp;

	/* Make sure module is disabled */
	vme64_out08( vmtgCSRB, VME64_CSR_OFF_BCLR, VME64_CSR_BIT_MODENBL );

	/* Program ADER                 */
	if ( vme64SetupADER( vmtgCSRB, VMTG_FUN, (uint32_t)vme_24_bus, VME_AM_STD_SUP_DATA ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to program address decoders for function 0\n");
		goto bail;
	}

	if ( 0 == devBusMappedRegister( "VMTG0", vmtgBase ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to register base address with devBusMapped\n");
		goto bail;
	}

	if ( 0 != devBusMappedRegisterIO( "vmtg32IO", &vmtg32IO ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to register vmtg32IO with devBusMapped\n");
		goto bail;
	}

	if ( 0 != devBusMappedRegisterIO( "vmtgCSR2Lo", &vmtgCSR2Lo ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to register vmtgCSR2Lo with devBusMapped\n");
		goto bail;
	}

	if ( 0 != devBusMappedRegisterIO( "vmtgCSR2Hi", &vmtgCSR2Hi ) ) {
		epicsPrintf("vmtgInit FAILED: Unable to register vmtgCSR2Hi with devBusMapped\n");
		goto bail;
	}

	vme64_out08( vmtgCSRB, VME64_CSR_OFF_BSET, VME64_CSR_BIT_MODENBL );
}/*end ifvmtg_flag*/
	rval     = 0;

bail:

	return rval;
}


long vmtgConfigure(void)
{
    static uint32_t already  = 0;

    if(already) {
        printf("The vmtgInit() has been called already\n");
        return -1;
    }

    already = 1;
    return vmtgInit();
}


static drvet drvVMTG = {
	number:    2,
	report:    vmtgReport,
	init:      vmtgInit
};

epicsExportAddress( drvet, drvVMTG );

/*Set Method for the theVmtgFlag used by the ioc */
void
setVmtg_flag(int flag){
        if ((flag==0) || (flag==1)){
                theVmtgFlag=flag;
        }
        else{ /*The expectation is that is called only once at startup */
                printf("setVmtg_flag: illegal value %i for VMTG flag. Value must be 0 or 1.",flag);
        }
}
/*Get Method for the theVmtgFlag used by the ioc */
int
getVmtg_flag(void){
        return   theVmtgFlag;
}

